### PR TITLE
refactor(service): add ForksafeService class

### PR DESCRIPTION
**Pull request by jd**
This adds a new class `ForksafeService` which is able to handle the lifecycle of
Service with respect to its processes (fork, exit, etc).

### CODE REVIEW

File: ddtrace/internal/service.py [hunk 1]
```diff
@@ -48,9 +62,94 @@ def _start(self):
...
+    ):
+        # type: (...) -> typing.Optional[float]
         """Join the service once stopped."""
+
+    def copy(self):
+        # type: () -> Service
+        return attr.evolve(self)
+
+
+class ServiceProcess(object):
```
Comment (P403n1x87) [inline ddtrace/internal/service.py:81]: I'm not quite happy with this name as, technically, this is not a process. I would suggest renaming it to something like `ForksafeService`

Comment (jd) [reply ddtrace/internal/service.py:81]: Sure I can rename it. I'll wait a bit in case other folks have different name ideas. 😄

Comment (Kyle-Verhoog) [reply ddtrace/internal/service.py:85]: I don't have a strong opinion. ForksafeService is pretty explicit which is good :)

File: tests/tracer/test_service.py [hunk 2]
```diff
@@ -25,3 +28,209 @@ def _start(self):
...
+
+    sm = MyServiceProcess()
+    sm.start()
+    assert sm.status == sm._service.status == service.ServiceStatus.RUNNING
+
+    sm.stop()
+    assert sm.status == sm._service.status == service.ServiceStatus.STOPPED
+
+
+def call_program(*args):
```
Comment (Kyle-Verhoog) [inline tests/tracer/test_service.py:65]: Seems like we should have this as a test util if we don't already 🙂

File: ddtrace/internal/service.py [hunk 3]
```diff
@@ -50,9 +66,89 @@ def _start(self):
...
+        self,
+        k,  # type: typing.Any
+    ):
+        # type: (...) -> typing.Any
+        return getattr(self._service, k)
+
+    def _restart_on_fork(self):
+        # type: () -> None
+        # Be sure to stop the parent first, since it might have to e.g. unpatch functions
+        # Do not flush data as we don't want to have multiple copies of the parent profile exported.
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/service.py:151]:
```suggestion
```

**Review by Kyle-Verhoog — COMMENTED** [review]
Just a few nits, otherwise looks great! 🚀

File: tests/tracer/test_service.py (hunk 2 above)
Comment (jd) [reply tests/tracer/test_service.py:65]: I thought about it but every test suite out there has its own version with its own use/edge case. I was not excited by the idea to try to write a one-solve-them-all function for all of those.

File: ddtrace/internal/service.py [hunk 4]
```diff
@@ -50,9 +66,89 @@ def _start(self):
...
+            atexit.register(self.stop)
+
+        if start_in_children:
+            forksafe.register(self._restart_on_fork)
+
+    def stop(self, *args, **kwargs):
+        # type: (...) -> None
+        """Stop the service."""
+        # Check for the service status: this *should* not be necessary per-say, but in the case `atexit.unregister` is
+        # unavailable, we're sure not to call stop twice.
```
[the remark this answers is not in the document]
Comment (jd) [reply ddtrace/internal/service.py:127]: if `unregister` is not there, then you can't `unregister` the `atexit` hook when `stop` is called so you might stop twice :)

File: tests/tracer/test_service.py (hunk 2 above)
Comment (Kyle-Verhoog) [reply tests/tracer/test_service.py:65]: fair enough :+1:

File: ddtrace/internal/service.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kyle-Verhoog) [reply ddtrace/internal/service.py:127]: Ohh I see lol I had assumed that if you couldn't unregister then you couldn't register either 😆

**Review by Kyle-Verhoog — DISMISSED** [review]

**Review by P403n1x87 — APPROVED** [review]

**Review by Kyle-Verhoog — DISMISSED** [review]

File: ddtrace/internal/periodic.py [hunk 5]
```diff
@@ -184,8 +184,12 @@ def interval(self, value):
         if self._worker:
             self._worker.interval = value
 
-    def _start(self):
-        # type: () -> None
+    def _start(
+        self,
+        *args,  # type: typing.Any
```
Comment (P403n1x87) [inline ddtrace/internal/periodic.py:189]: Is it not possible to have a more precise signature for these methods?

Comment (jd) [reply ddtrace/internal/periodic.py:189]: Not really as it's just passing the plate to the inheriting class.
And since the parent classes are accepting any arguments for `start` to allow inheriting class to get those, it makes it impossible to set a more specific type there.

**Comment by jd** [comment]
This should get some tests to cover a bit better issues like #2295

**Comment by jd** [comment]
Added some more tests

File: ddtrace/internal/service.py [hunk 6]
```diff
@@ -56,5 +76,87 @@ def stop(self):
...
+
+        :param stop_on_exit: Whether to stop the service and wait for shutdown on exit.
+        :param start_in_children: Whether to start the service in child processes.
+        """
+        stop_on_exit = kwargs.pop("stop_on_exit", True)
+        start_in_children = kwargs.pop("start_in_children", True)
+
+        if start_in_children:
+            try:
+                uwsgi.check_uwsgi(self._restart_on_fork, atexit=self.stop if stop_on_exit else None)
```
Comment (majorgreys) [inline ddtrace/internal/service.py:119]: This would make it so that uwsgi child processes restart twice as you are registering both post fork hooks with uwsgi and with `os.register_at_fork` which will be called once `PyOS_AfterFork[_Child]` is called.

Comment (jd) [reply ddtrace/internal/service.py:119]: uwsgi calls the `fork()` system calls, not Python's fork. So those hooks are not called.
Otherwise, we wouldn't need a custom uwsgi code at all. 🤷‍♂️

Comment (majorgreys) [reply ddtrace/internal/service.py:119]: After calling the system `fork()`, uwsgi does end up calling PyOS_AfterFork_Child which in turn will call the hooks that had been registered:

* https://github.com/unbit/uwsgi/search?q=PyOS_AfterFork
* https://github.com/python/cpython/blob/95e4431804587a0c9d464bb7b3d5f3057bbeaccd/Modules/posixmodule.c#L588
* https://github.com/python/cpython/blob/95e4431804587a0c9d464bb7b3d5f3057bbeaccd/Modules/posixmodule.c#L6566-L6595

Comment (majorgreys) [reply ddtrace/internal/service.py:119]: The difference in uWSGI is that by default `PyOS_AfterFork_Child` is not called early on in the child process after it is forked. It can be called earlier if the `py-call-osafterfork` option is set.

* https://github.com/unbit/uwsgi/blob/d58a832c81c2c96ae0f6e72614e1cc47f4b5d332/plugins/python/python_plugin.c#L435-L442
* https://github.com/unbit/uwsgi/blob/d58a832c81c2c96ae0f6e72614e1cc47f4b5d332/plugins/python/python_plugin.c#L2048-L2053

Comment (jd) [reply ddtrace/internal/service.py:119]: `PyOS_AfterFork_Child` is actually never called unless you enable `py-call-osafterfork` IIUC. So I think that code is still right, *except* if `py-call-osafterfork` is enabled. 😞 

That might be something we need to detect and check, even in the current `ddtrace.internal.uwsgi` module 😢

Comment (majorgreys) [reply ddtrace/internal/service.py:119]: Sorry, I misunderstood how uwsgi was doing things. `PyOS_AfterFork_Child` is called in [`uwsgi_python_post_fork`](https://github.com/unbit/uwsgi/blob/d58a832c81c2c96ae0f6e72614e1cc47f4b5d332/plugins/python/python_plugin.c#L435-L442) only if `py-call-osafterfork` is enabled. If it is disabled `PyOS_AfterFork_Child` it can still be called in [`uwsgi_python_worker`](https://github.com/unbit/uwsgi/blob/d58a832c81c2c96ae0f6e72614e1cc47f4b5d332/plugins/python/python_plugin.c#L2048-L2053) but only if you are using [worker overrides](https://uwsgi-docs.readthedocs.io/en/latest/WorkerOverride.html).

Going back to the initial comment, it would still seem like we don't want to register hooks in uwsgi and in Python if we know we are running within uwsgi. Whether or not the user has enabled `py-call-osafterfork` would not matter here as the uwsgi post fork hooks would be [called](https://github.com/unbit/uwsgi/blob/d58a832c81c2c96ae0f6e72614e1cc47f4b5d332/plugins/python/python_plugin.c#L446-L454) regardless.

As long as a uwsgi user is not using `ddtrace-run`, they could still start such `ForksafeServices` in the master process, register hooks with `ddtrace.internal.uwsgi`, and have those services restarted in the child worker processes.

Comment (jd) [reply ddtrace/internal/service.py:119]: > Going back to the initial comment, it would still seem like we don't want to register hooks in uwsgi and in Python if we know we are running within uwsgi. Whether or not the user has enabled py-call-osafterfork would not matter here as the uwsgi post fork hooks would be called regardless.

This code does not register after fork hook in the master process, it only does in the children process. Which is correct, since it's possible for a uwsgi child process to call `os.fork`.

If the user enables `py-call-osafterfork` then the problem also does not arise: the `start` method does not register `forksafe` hook in the master process, it just returns right away and wait for the uwsgi postfork hook to be triggered – so we're still fine.

WDYT?

Comment (majorgreys) [reply ddtrace/internal/service.py:119]: Ok, sorry it took me a few back and forths to understand what's happening in the case of uwsgi. The restart will be registered as uwsgi post fork hook in the master but not with forksafe, in the child the restart will be registered with forksafe so if a worker calls `os.fork` then it'll sure to be started. 

Finally, `py-call-osafterfork` does not matter to this functionality as these restarts are handled in the master process by uwsgi's own post fork hooks.

Comment (jd) [reply ddtrace/internal/service.py:119]: Exactly @majorgreys!

… so why don't I have a ✅  ? 😁

File: ddtrace/internal/service.py [hunk 7]
```diff
@@ -56,5 +76,87 @@ def stop(self):
...
+
+        if start_in_children:
+            forksafe.register(self._restart_on_fork)
+
+    def stop(self):
+        # type: () -> None
+        """Stop the service."""
+        # Check for the service status: this *should* not be necessary per-say, but in the case `atexit.unregister` is
+        # unavailable, we're sure not to call stop twice.
+        if self._service.status == ServiceStatus.RUNNING:
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/service.py:137]: Is there a possible race condition here? We handle starting race conditions but seems like it'd also make sense to do it on stop() too (unless we state that stop() may be called more than once)

**Review by Kyle-Verhoog — COMMENTED** [review]
Just one thought/concern from me, can't really speak to all the uwsgi stuff 😆 

Otherwise looks good!

File: ddtrace/internal/service.py (hunk 7 above)
Comment (jd) [reply ddtrace/internal/service.py:137]: We don't handle race conditions in `start` and this whole class is **not** thread-safe.

Does it have to be thread-safe?

In the case of the `Profiler`, I'm OK stating that the `Profiler` object is not thread-safe and that you cannot call `start()` multiple times from different threads.

Comment (Kyle-Verhoog) [reply ddtrace/internal/service.py:137]: Oh yeah I meant we handle the underlying start race condition in `Service` but we don't with stop.

Yeah I don't think it has to be thread-safe but wanted to make sure that we were aware of it - maybe worth indicating it explicitly in the docstring?

> In the case of the Profiler, I'm OK stating that the Profiler object is not thread-safe and that you cannot call start() multiple times from different threads.

:+1: agreed

Comment (jd) [reply ddtrace/internal/service.py:137]: Sure, I've added something in the doc string.

**Comment by jd** [comment]
I've refactored this a it and it led to https://github.com/DataDog/dd-trace-py/pull/2450
Will rebase once its merged.
